sys/net/link_layer/ieee802154/submac.c

```diff
@@ -431,6 +431,208 @@ int ieee802154_send(ieee802154_submac_t *submac, const iolist_t *iolist)
+         + IEEE802154G_ATURNAROUNDTIME_US;
+}
+
+/*
+ * MR-OFDM timing calculations
+ *
+ * The standard unfortunately does not list the formula, instead it has to be pieced together
+ * from scattered information and tables in the IEEE 802.15.4 document - may contain errors.
+ */
+
+static unsigned _mr_ofdm_frame_duration(uint8_t option, uint8_t scheme, uint8_t bytes)
+{
+    /* Table 150 - phySymbolsPerOctet values for MR-OFDM PHY, IEEE 802.15.4g-2012 */
+    static const uint8_t quot[] = { 3, 3, 6, 12, 18, 24, 36 };
+
+    --option;
+    /* phyMaxFrameDuration = phySHRDuration + phyPHRDuration + ceiling [(aMaxPHYPacketSize + 1) x phySymbolsPerOctet] */
+    const unsigned phySHRDuration = 6;
+    const unsigned phyPHRDuration = option ? 6 : 3;
+    const unsigned phyPDUDuration = ((bytes + 1) * (1 << option) + quot[scheme] - 1)
+                                  / quot[scheme];
+
+    return (phySHRDuration + phyPHRDuration + phyPDUDuration) * IEEE802154_MR_OFDM_SYMBOL_TIME_US;
+}
+
+static inline uint16_t _mr_ofdm_csma_backoff_period_us(const ieee802154_mr_odmf_conf_t *conf)
+{
+    (void)conf;
+
+    return IEEE802154_CCA_DURATION_IN_SYMBOLS * IEEE802154_MR_OFDM_SYMBOL_TIME_US
+         + IEEE802154G_ATURNAROUNDTIME_US;
+}
+
+MAYBE_UNUSED
+static inline uint16_t _mr_ofdm_ack_timeout_us(const ieee802154_mr_odmf_conf_t *conf)
+{
+    return _mr_ofdm_csma_backoff_period_us(conf)
+         + IEEE802154G_ATURNAROUNDTIME_US
+         + _mr_ofdm_frame_duration(conf->option, conf->scheme, IEEE802154_ACK_FRAME_LEN);
+}
+
+/*
+ * MR-FSK timing calculations
+ *
+ * The standard unfortunately does not list the formula, instead it has to be pieced together
+ * from scattered information and tables in the IEEE 802.15.4 document - may contain errors.
+ */
+
+MAYBE_UNUSED
+static inline uint16_t _mr_fsk_csma_backoff_period_us(const ieee802154_mr_fsk_conf_t *conf)
+{
+    (void)conf;
+
+    return IEEE802154_CCA_DURATION_IN_SYMBOLS * IEEE802154_MR_FSK_SYMBOL_TIME_US
+         + IEEE802154G_ATURNAROUNDTIME_US;
+}
+
+MAYBE_UNUSED
+static inline uint16_t _mr_fsk_ack_timeout_us(const ieee802154_mr_fsk_conf_t *conf)
+{
+    uint8_t ack_len = IEEE802154_ACK_FRAME_LEN;
+    uint8_t fsk_pl = ieee802154_mr_fsk_plen(conf->srate);
+
+    /* PHR uses same data rate as PSDU */
+    ack_len += 2;
+
+    /* 4-FSK doubles data rate */
+    if (conf->mod_ord == 4) {
+        ack_len /= 2;
+    }
+
+    /* forward error correction halves data rate */
+    if (conf->fec) {
+        ack_len *= 2;
+    }
+
+    return _mr_fsk_csma_backoff_period_us(conf)
+         + IEEE802154G_ATURNAROUNDTIME_US
+         /* long Preamble + SFD; SFD=2 */
+         + ((fsk_pl * 8 + 2) + ack_len) * 8 * IEEE802154_MR_FSK_SYMBOL_TIME_US;
+}
+
+static int ieee802154_submac_config_phy(ieee802154_submac_t *submac,
+                                        const ieee802154_phy_conf_t *conf)
+{
+    switch (conf->phy_mode) {
+    case IEEE802154_PHY_OQPSK:
```

Comment:
> alternative to https://github.com/RIOT-OS/RIOT/pull/19198

`IEEE802154_PHY_BPSK` and `IEEE802154_PHY_ASK` are not handled here so far.

I assume what is missing is a mixture of this implementation style and the symbol times from #19198

Reply:
[ieee_phy_ack_time.patch.txt](https://github.com/user-attachments/files/18069262/ieee_phy_ack_time.patch.txt)

I took #19198 and compared it with the corresponding sections in the standards.
I think I implemented the ACK timeouts for OQPSK, BPSK and ASK.

[This](https://onlinedocs.microchip.com/oxy/GUID-4B3D771E-5647-4191-AD45-C897B0D23071-en-US-3/GUID-D1CB35D8-C789-48E3-BA22-713F04E39553.html) and [This](https://groups.google.com/g/omnetpp/c/q0hgsnVvJ0I?pli=1) did also help me to understand the symbol and timing relations